scio-extra/src/main/scala/com/spotify/scio/extra/hll/zetasketch/package.scala

```diff
@@ -0,0 +1,171 @@
+import com.spotify.scio.util.TupleFunctions._
+import com.spotify.scio.values.SCollection
+import org.apache.beam.sdk.extensions.zetasketch.HllCount
+
+package object zetasketch {
+
+  /**
+   * [[com.spotify.scio.estimators.ApproxDistinctCounter]] implementation for
+   * [[org.apache.beam.sdk.extensions.zetasketch.HllCount]],
+   * specifically [[org.apache.beam.sdk.extensions.zetasketch.HllCount.Init#forIntegers]].
+   * HllCount estimate the distinct count using HyperLogLogPlusPlus (HLL++) sketches on data streams based on
+   * the ZetaSketch implementation.
+   *
+   * The HyperLogLog++ (HLL++) algorithm estimates the number of distinct values in a data stream.
+   * HLL++ is based on HyperLogLog; HLL++ more accurately estimates the number of distinct values in very large and
+   * small data streams.
+   *
+   * @param p Precision, controls the accuracy of the estimation. The precision value will have an impact on the number of buckets
+   *          used to store information about the distinct elements.
+   *          should be in the range `[10, 24]`, default precision value is `15`.
+   */
+  case class ZetasketchHllIntCounter(p: Int = HllCount.DEFAULT_PRECISION)
+      extends ApproxDistinctCounter[Int] {
+
+    override def estimateDistinctCount(in: SCollection[Int]): SCollection[Long] = {
+      in.asInstanceOf[SCollection[Integer]]
+        .applyTransform(HllCount.Init.forIntegers().withPrecision(p).globally())
+        .applyTransform(HllCount.Extract.globally())
+        .map(Long2long)
+    }
+
+    override def estimateDistinctCountPerKey[K](
+      in: SCollection[(K, Int)]
+    ): SCollection[(K, Long)] = {
+      implicit val (keyCoder, _): (Coder[K], Coder[Int]) = BeamCoders.getTupleCoders(in)
+      in.mapValues(int2Integer)
+        .toKV
+        .applyTransform(HllCount.Init.forIntegers().withPrecision(p).perKey())
+        .applyTransform(HllCount.Extract.perKey())
+        .map(klToTuple)
+    }
+  }
+
+  /**
+   * [[com.spotify.scio.estimators.ApproxDistinctCounter]] implementation for
+   * [[org.apache.beam.sdk.extensions.zetasketch.HllCount]],
+   * specifically [[org.apache.beam.sdk.extensions.zetasketch.HllCount.Init#forLongs]].
+   * HllCount estimate the distinct count using HyperLogLogPlusPlus (HLL++) sketches on data streams based on
+   * the ZetaSketch implementation.
+   *
+   * The HyperLogLog++ (HLL++) algorithm estimates the number of distinct values in a data stream.
+   * HLL++ is based on HyperLogLog; HLL++ more accurately estimates the number of distinct values in very large and
+   * small data streams.
+   *
+   * @param p Precision, controls the accuracy of the estimation. The precision value will have an impact on the number of buckets
+   *          used to store information about the distinct elements.
+   *          should be in the range `[10, 24]`, default precision value is `15`.
+   */
+  case class ZetasketchHllLongCounter(p: Int = HllCount.DEFAULT_PRECISION)
```

Comment:
You can probably dedup these for specializations with a type class like `ZetaSketchable[T]` and implicit impls for `Int/Long/String/Array[Byte]`?

Reply:
do you have an example where we do this type of abstraction, I am struggling to see the complete picture here and how to use implicit impls. 🤔

Reply:
You can unify the 4 case cases like this, basically abstract away the `HllCount.Init.for*()` method calls, because everything else is the same. This also allows us to support other types without adding a new case class.

```
case class ZetaSketchHLLCounter[T](...)(implicit zs: ZetaSketchable[T]) {
  def estimateDistinctCount(in: ...) = in
    .sth
    .applyTransform(zs.init)
    .sth
}

sealed trait ZetaSketchable[T] {
  def init: ...
}

object ZetaSketchable {
  implicit val intZetaSketchable = new ZetaSketchable[T] {
    def init: ... = HllCount.Init.forInts()
  }
}
```

And later in the follow up PR we can add:
```
sealed trait ZetaSketchable[T] {
  def decode(ba: Array[Byte]): ZetaHLL[T]
  def encode(zh: ZetaHLL[T]): Array[Byte]
}

sealed trait ZetaHLL[T] {
  def estimateSize: Long = ...
  def merge(that: ZetaHLL[T]): ZetaHLL[T]
}
```

That allows as to have methods like `SCollection[T]#asZetaHLL: SCollection[ZetaHLL[T]]` similar to the `ApproxFilter` abstractions.